File: user-service/src/main/java/socialapp/userservice/common/mapper/UserMapper.java
package socialapp.userservice.common.mapper;

import org.mapstruct.*;
import socialapp.userservice.model.dto.RegistrationDto;
import socialapp.userservice.model.dto.UserUpdateDto;
import socialapp.userservice.model.entity.Address;
import socialapp.userservice.model.entity.User;

@Mapper(
        nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE,
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
)
public interface UserMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "address", expression = "java(address)")
    User toModel(RegistrationDto registrationDto, Address address);

    void update(UserUpdateDto userUpdateDto, @MappingTarget User user);
}

File: kt-user-service/src/main/kotlin/socialapp/ktuserservice/common/mapper/UserMapper.kt
package socialapp.ktuserservice.common.mapper

import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingTarget
import org.mapstruct.NullValueCheckStrategy
import org.mapstruct.NullValuePropertyMappingStrategy
import socialapp.ktuserservice.model.dto.RegistrationDto
import socialapp.ktuserservice.model.dto.UserUpdateDto
import socialapp.ktuserservice.model.entity.Address
import socialapp.ktuserservice.model.entity.User

@Mapper(
    nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
)
interface UserMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "address", expression = "java(address)")
    fun toModel(registrationDto: RegistrationDto, address: Address): User

    fun update(userUpdateDto: UserUpdateDto, @MappingTarget user: User)
}

